contracts/governed_token_wrapper/lib.rs

```diff
@@ -0,0 +1,274 @@
+
+    /// The token wrapper error types.
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        /// Invalid amount provided for native wrapping
+        InvalidAmountForNativeWrapping,
+        /// Native wrapping is not allowed for this token wrapper
+        NativeWrappingNotAllowed,
+        /// Invalid value sent for wrapping
+        InvalidValueSentForWrapping,
+        /// Invalid token address
+        InvalidTokenAddress,
+        /// Invalid token amount
+        InvalidTokenAmount,
+        /// Insufficient native balance
+        InsufficientNativeBalance,
+        /// Native unwrapping is not allowed for this token wrapper
+        NativeUnwrappingNotAllowed,
+        /// Insufficient PSP22 balance
+        InsufficientPSP22Balance,
+        /// Invalid historical token address
+        InvalidHistoricalTokenAddress,
+    }
+
+    impl GovernedTokenWrapper {
+        #[ink(constructor)]
+        pub fn new(
+            name: Option<String>,
+            symbol: Option<String>,
+            decimal: u8,
+            governor: AccountId,
+            fee_recipient: AccountId,
+            fee_percentage: Balance,
+            is_native_allowed: bool,
+            wrapping_limit: u128,
+            proposal_nonce: u64,
+            token_address: AccountId,
+        ) -> Self {
+            ink_lang::codegen::initialize_contract(|instance: &mut Self| {
+                // for wrapping
+                instance._init(token_address);
+
+                instance.metadata.name = name;
+                instance.metadata.symbol = symbol;
+                instance.metadata.decimals = decimal;
+
+                // Governance config
+                instance.governor = governor;
+                instance.fee_recipient = fee_recipient;
+                instance.fee_percentage = fee_percentage;
+                instance.is_native_allowed = is_native_allowed;
+                instance.wrapping_limit = wrapping_limit;
+                instance.proposal_nonce = proposal_nonce;
+            })
+        }
+
+        /// Used to wrap tokens on behalf of a sender.
+        ///
+        /// token_address is the address of PSP22 to transfer to, if token_address is None,
+        /// then it's a Native token address
+        ///
+        /// amount is the amount of token to transfer
+        #[ink(message, payable)]
+        pub fn wrap(&mut self, token_address: Option<AccountId>, amount: Balance) {
+            self.is_valid_wrapping(token_address, amount);
+
+            // determine amount to use
+            let amount_to_use = if token_address.is_none() {
+                self.env().transferred_value()
+            } else {
+                amount
+            };
+
+            let cost_to_wrap = self.get_fee_from_amount(amount_to_use);
+
+            let leftover = amount_to_use.saturating_mul(cost_to_wrap);
```

Comment:
Isn't this `saturating_sub` or anything like that? instead of `saturating_mul`.

Reply:
yh. should be